### PR TITLE
NAS-129579 / 25.04 / Add custom app form

### PR DESCRIPTION
**Pull request by RehanY147**
**Changes:**

Adds a form to install apps using custom YML config

**Testing:**

Go to the apps discover page, and click on the `'Custom App'` button. Enter a unique app name and enter this config in the `"Custom Config"` input.
```
services:
  nginx:
    image: nginx:latest
    ports:
      - "8580:80"
    restart: unless-stopped
```

### Downstream

|Affects         |Reasoning
|----------------|-------------------------------
|Documentation   | Adds a form to install custom app with YAML config

### CODE REVIEW

File: src/app/pages/apps/components/custom-app-form/custom-app-form.component.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    private cdr: ChangeDetectorRef,
+    private ws: WebSocketService,
+    private errorHandler: ErrorHandlerService,
+    private dialogService: DialogService,
+    private appService: ApplicationsService,
+    private dialogRef: IxSlideInRef<CustomAppFormComponent>,
+  ) {}
+
+  ngOnInit(): void {
+    this.addForbiddenAppNamesValidator();
```
Comment (bvasilenko) [inline src/app/pages/apps/components/custom-app-form/custom-app-form.component.ts:48]: I've noticed that the form could not pass validation when there's no installed apps. I'd double check the validation logic.

**Review by bvasilenko — CHANGES_REQUESTED** [review]

File: src/app/pages/apps/components/custom-app-form/custom-app-form.component.scss
Comment (undsoft) [inline src/app/pages/apps/components/custom-app-form/custom-app-form.component.scss:1]: Empty file can removed.

File: src/app/pages/apps/components/available-apps/custom-app-button/custom-app-button.component.html [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <!-- TODO: https://ixsystems.atlassian.net/browse/NAS-129579 -->
-@if (false) {
+<!-- @if (false) { -->
```
Comment (undsoft) [inline src/app/pages/apps/components/available-apps/custom-app-button/custom-app-button.component.html:2]: You can just remove this.

File: src/app/pages/apps/components/custom-app-form/custom-app-form.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    ],
+  });
+
+  beforeEach(() => {
+    spectator = createComponent();
+    loader = TestbedHarnessEnvironment.loader(spectator.fixture);
+  });
+
+  it('closes slide in when successfully submitted', async () => {
+    spectator.component.form.patchValue({
```
Comment (undsoft) [inline src/app/pages/apps/components/custom-app-form/custom-app-form.component.spec.ts:72]: When writing tests it's best to cover public interfaces of whatever you are testing. In this case `form` is an implementation detail of the component. It's not mean to be interacted with directly. It should be `protected` really. You should test form by interacting with controls.

**Review by undsoft — COMMENTED** [review]
Looks good in general. Can we have code-editor have at least 3 lines by default?

**Comment by RehanY147** [comment]
@bvasilenko Can you provide steps to reproduce the validation issue? It's working for me.

**Review by undsoft — APPROVED** [review]
Looks good.

**Comment by bvasilenko** [comment]
@RehanY147 curerntly I cannot reproduce the issue any more

**Review by bvasilenko — APPROVED** [review]
Works good!

**Comment by undsoft** [comment]
backport

**Pull request merged**